src/server/editorServices.ts

```diff
@@ -259,67 +245,60 @@ namespace ts.server {
+        MissingFilePath = "Missing file from program",
+        WildCardDirectories = "Wild card directory",
+        TypeRoot = "Type root of the project",
+        ClosedScriptInfo = "Closed Script info",
+        ConfigFileForInferredRoot = "Config file for the inferred project root"
     }
 
-    /**
-     * TODO: enforce invariants:
-     *  - script info can be never migrate to state - root file in inferred project, this is only a starting point
-     *  - if script info has more that one containing projects - it is not a root file in inferred project because:
-     *    - references in inferred project supercede the root part
-     *    - root/reference in non-inferred project beats root in inferred project
-     */
-    function isRootFileInInferredProject(info: ScriptInfo): boolean {
-        if (info.containingProjects.length === 0) {
-            return false;
-        }
-        return info.containingProjects[0].projectKind === ProjectKind.Inferred && info.containingProjects[0].isRoot(info);
+    /* @internal */
+    export const enum WatcherCloseReason {
+        ProjectClose = "Project close",
+        NotNeeded = "After project update isnt required any more",
+        FileCreated = "File got created",
+        RecursiveChanged = "Recursive changed for the watch",
+        ProjectReloadHitMaxSize = "Project reloaded and hit the max file size capacity",
+        OrphanScriptInfoWithChange = "Orphan script info, Detected change in file thats not needed any more",
+        OrphanScriptInfo = "Removing Orphan script info as part of cleanup",
+        FileDeleted = "File was deleted",
+        FileOpened = "File opened",
+        ConfigProjectCreated = "Config file project created",
+        FileClosed = "File is closed"
+    }
+
+    const enum ConfigFileWatcherStatus {
+        ReloadingFiles = "Reloading configured projects for files",
+        ReloadingInferredRootFiles = "Reloading configured projects for only inferred root files",
+        UpdatedCallback = "Updated the callback",
+        TrackingFileAdded = "Tracking file added",
+        TrackingFileRemoved = "Tracking file removed",
+        InferredRootAdded = "Inferred Root file added",
+        InferredRootRemoved = "Inferred Root file removed",
     }
 
-    class DirectoryWatchers {
+    /* @internal */
+    export type ServerDirectoryWatcherCallback = (path: NormalizedPath) => void;
+
+    type ConfigFileExistence = {
         /**
-         * a path to directory watcher map that detects added tsconfig files
+         * Cached value of existence of config file
          */
-        private readonly directoryWatchersForTsconfig: Map<FileWatcher> = createMap<FileWatcher>();
+        exists: boolean;
         /**
-         * count of how many projects are using the directory watcher.
-         * If the number becomes 0 for a watcher, then we should close it.
+         * The value in the open files map is true if the file is inferred project root

```

Comment:
File is *an* inferred project root or *the* inferred project root?
If *the*, this should probably be just store the root instead of a `Map` with 0 or 1 entries.